Fix ManyToOneType.IsModified to handle both object instance and identifier passed to the "old" parameter

mishaurubkov opened the pull request:
… identifier passed to the parameter “old”.

The tests `EventListener_Entity_ChangeProperty` and `EventListener_EntityWithCompositeId_ChangeProperty` fail for current code in `ManyToOneType.IsModified(…)` where `old` parameter is not parsed.

The tests `SelectBeforeUpdate_EntityWithCompositeId_ChangeProperty` and `SelectBeforeUpdate_ManyToOneWithCompositeId_ChangeProperty` fail
 if we apply Hibernate fix [HHH-3730](https://hibernate.atlassian.net/browse/HHH-3730) `(var oldIdentifier =GetIdentifier(old, session))`

All tests pass if we apply proposed fix: `var oldIdentifier = IsIdentifier(old, session) ? old : GetIdentifier(old, session);`

I added extra tests to see what data hit the `ManyToOneType.IsModified(…)`. We can leave just 4 tests listed above and remove or set to `[Ignore]` others.

`ManyToOneType.IsModifiedAsync(...)` could be fixed same way but would like to hear the feedback on the issue and proposed solution first.

Fixes #1496.

fredericDelaporte commented:
> ManyToOneType.IsModifiedAsync(...) could be fixed same way but would like to hear the feedback on the issue and proposed solution first.

No need to wait, async code is generated: just regenerate it. (Do not manually edit it.) Follow [contributing](https://github.com/nhibernate/nhibernate-core/blob/master/CONTRIBUTING.md) instructions. [Async generation](https://github.com/nhibernate/nhibernate-core/blob/master/CONTRIBUTING.md#regenerate-async-code) is an option of the [build menu](https://github.com/nhibernate/nhibernate-core/blob/master/CONTRIBUTING.md#the-build-menu).

mishaurubkov commented:
DONE. Generated Async method with tests. See #98b1918 commit.

fredericDelaporte commented on src/NHibernate.Test/NHSpecificTest/GH1496/AuditEventListener.cs:
```diff
@@ -0,0 +1,69 @@
+			isActive = false;
+		}
+
+		public async Task OnPostUpdateAsync(PostUpdateEvent @event, CancellationToken cancellationToken)
```
Normally there is a no need to write this async method too. The async generator will do it for you too. Of course this means that at first your class will lack some interface method implementations and the project will not compile till you run the async generator, but this is the preferred way.

fredericDelaporte commented on src/NHibernate.Test/NHSpecificTest/GH1496/Entities.cs:
```diff
@@ -0,0 +1,129 @@
+		}
+	}
+
+	// Entity with ManyToOne type (Contact) with comnposite (ContactId) Id.
```
Typo on `comnposite`.

fredericDelaporte commented on src/NHibernate.Test/NHSpecificTest/GH1496/Entities.cs:
```diff
@@ -0,0 +1,129 @@
+	}
+
+
+	public class ContactIdentifier : IEquatable<ContactIdentifier>
```
Shouldn't this be marked serializable too?

fredericDelaporte commented on src/NHibernate.Test/NHSpecificTest/GH1496/Entities.cs:
```diff
@@ -0,0 +1,129 @@
+	}
+
+
+
```
Please clean the file from such consecutive blank lines.

mishaurubkov commented:
I did cleanup and committed the changes:
1) OnPostUpdateAsync was created using generator
2) Removed spaces
3) I didn't really need [Serializable] so I removed it.
